File: code/dart/modular_api_rest_client/lib/src/client_tracing.dart
import 'package:dartastic_opentelemetry_api/dartastic_opentelemetry_api.dart';

/// The header carrying a caller-supplied correlation id.
const String requestIdHeader = 'x-request-id';

/// Instrumentation for one outbound call.
///
/// **Zero configuration on purpose.** The parent comes from the ambient context — the
/// server span the host made active — and the tracer comes from the global provider.
/// When no OpenTelemetry SDK is installed both are no-ops, so a consumer who never
/// enables tracing gets no spans, no headers and no measurable cost. Nothing has to be
/// wired into `ServiceClientConfig`.
///
/// **Injection goes through the global propagator**, never through a propagator of our
/// own. This package does not depend on `modular_api` (runbook D21), so it cannot reach
/// the framework's W3C propagator — and it should not: the OpenTelemetry API's own
/// documentation says instrumentation libraries should *read* the global propagator and
/// leave setting it to the SDK or the host. That also means whatever chain the
/// application configured is honoured here, including a Cloud Trace propagator it
/// appended, without this package knowing such a thing exists.
class ClientCallTracing {
  ClientCallTracing._(this._span);

  final APISpan? _span;

  /// Starts a client span for [operationId] against [uri], and injects trace context
  /// plus the caller's request id into [headers].
  ///
  /// Returns a handle whose [complete] must be called exactly once.
  static ClientCallTracing start({
    required String method,
    required String operationId,
    required Uri uri,
    required Map<String, String> headers,
    String? inboundRequestId,
  }) {
    // Forwarded whether or not tracing is on. Envoy's tracing documentation lists
    // x-request-id propagation as a separate obligation from trace context, and it is the
    // one correlation header nearly every stack implements — so it survives a hop into a
    // service that has no tracing at all (runbook D23). We forward, never invent: a chain
    // that has no request id keeps having none.
    if (inboundRequestId != null && inboundRequestId.isNotEmpty) {
      headers.putIfAbsent(requestIdHeader, () => inboundRequestId);
    }

    final parent = Context.current;
    final parentSpan = parent.span;

    // No recording parent means tracing is off for this request. Creating a
    // non-recording span and injecting a header nobody will read would be cost with no
    // benefit.
    if (parentSpan == null || !parentSpan.spanContext.isValid) {
      return ClientCallTracing._(null);
    }

    final span = OTelAPI.tracerProvider().getTracer('modular_api_rest_client').startSpan(
      // Semantic convention for a client span: the method, qualified by what was called.
      // The operation id is more useful than a bare method and safer than a full URL,
      // which can carry identifiers in its path.
      '$method $operationId',
      context: parent,
      kind: SpanKind.client,
      attributes: OTelAPI.attributesFromMap(<String, Object>{
        'http.request.method': method,
        'server.address': uri.host,
        if (uri.hasPort) 'server.port': uri.port,
        'url.path': uri.path,
      }),
    );

    OTelAPI.textMapPropagator.inject(
      parent.withSpanContext(span.spanContext),
      headers,
      _MapSetter(headers),
    );

    return ClientCallTracing._(span);
  }

  /// Ends the span, recording the outcome.
  ///
  /// [statusCode] is null when the call never produced a response — a timeout, a DNS
  /// failure, a refused connection. That case is always an error; a status code is only
  /// an error at 4xx and above, because a client span's failure is the *call* failing,
  /// not the server disagreeing.
  void complete({int? statusCode, Object? error}) {
    final span = _span;
    if (span == null) return;

    if (statusCode != null) {
      span.setIntAttribute('http.response.status_code', statusCode);
    }

    if (error != null) {
      span.recordException(error);
      span.setStatus(SpanStatusCode.Error, error.runtimeType.toString());
    } else if (statusCode == null) {
      span.setStatus(SpanStatusCode.Error, 'no response');
    } else if (statusCode >= 400) {
      // Unlike a server span, where 4xx is the caller's mistake, a 4xx here means *our*
      // outbound call failed. The distinction matters when reading a waterfall.
      span.setStatus(SpanStatusCode.Error, 'HTTP $statusCode');
    }

    span.end();
  }
}

class _MapSetter extends TextMapSetter<String> {
  _MapSetter(this._headers);

  final Map<String, String> _headers;

  @override
  void set(String key, String value) => _headers[key] = value;
}

File: code/ts/modular_api_rest_client/src/clientTracing.ts
import {
  context as otelContext,
  propagation,
  SpanKind,
  SpanStatusCode,
  trace,
  type Span,
} from '@opentelemetry/api';

/** The header carrying a caller-supplied correlation id. */
export const REQUEST_ID_HEADER = 'x-request-id';

export interface ClientCallTracingInit {
  readonly method: string;
  readonly operationId: string;
  readonly url: string | URL;
  readonly headers: Record<string, string>;
}

/**
 * Instrumentation for one outbound call.
 *
 * **Zero configuration on purpose.** The parent comes from the ambient context — the
 * server span the host made active — and the tracer from the global provider. When no
 * OpenTelemetry SDK is registered both are no-ops, so a consumer who never enables tracing
 * gets no spans, no headers and no measurable cost. Nothing is threaded into
 * `ServiceClientConfig`.
 *
 * **Injection goes through the global propagator**, never a propagator of our own. This
 * package does not depend on `@macss/modular-api` (runbook D21), so it cannot reach the
 * framework's W3C propagator — and per the OpenTelemetry API's own guidance it should not:
 * instrumentation libraries read the global and leave setting it to the SDK or host. That
 * also means whatever chain the application configured is honoured here, including a Cloud
 * Trace propagator it appended, without this package knowing such a thing exists.
 *
 * The Dart counterpart is `lib/src/client_tracing.dart`.
 */
export class ClientCallTracing {
  private constructor(private readonly span: Span | undefined) {}

  /**
   * Starts a client span and injects trace context plus the caller's request id into
   * `headers`.
   */
  static start(init: ClientCallTracingInit, inboundRequestId?: string): ClientCallTracing {
    // Forwarded whether or not tracing is on. Envoy's tracing documentation lists
    // x-request-id propagation as an obligation separate from trace context, and it is the
    // one correlation header nearly every stack implements — so it survives a hop into a
    // service with no tracing at all (runbook D23). We forward, never invent: a chain with
    // no request id keeps having none.
    if (inboundRequestId !== undefined && inboundRequestId.length > 0) {
      init.headers[REQUEST_ID_HEADER] ??= inboundRequestId;
    }

    const parent = otelContext.active();
    const parentSpan = trace.getSpan(parent);

    // No recording parent means tracing is off for this request. A non-recording span and a
    // header nobody will read would be cost with no benefit.
    if (parentSpan === undefined || !parentSpan.isRecording()) {
      return new ClientCallTracing(undefined);
    }

    const target = safeUrl(init.url);

    const span = trace.getTracer('modular_api_rest_client').startSpan(
      // Semantic convention for a client span: the method, qualified by what was called.
      // The operation id is more useful than a bare method and safer than a full URL, which
      // can carry identifiers in its path.
      `${init.method} ${init.operationId}`,
      {
        kind: SpanKind.CLIENT,
        attributes: {
          'http.request.method': init.method,
          ...(target === undefined
            ? {}
            : {
                'server.address': target.hostname,
                ...(target.port === '' ? {} : { 'server.port': Number(target.port) }),
                'url.path': target.pathname,
              }),
        },
      },
      parent,
    );

    propagation.inject(trace.setSpan(parent, span), init.headers);

    return new ClientCallTracing(span);
  }

  /**
   * Ends the span, recording the outcome.
   *
   * `statusCode` is undefined when the call never produced a response — a timeout, a DNS
   * failure, a refused connection. That is always an error; a status code is an error at
   * 4xx and above, because a client span's failure is the *call* failing rather than the
   * server disagreeing.
   */
  complete({ statusCode, error }: { statusCode?: number; error?: unknown } = {}): void {
    const { span } = this;
    if (span === undefined) return;

    if (statusCode !== undefined) {
      span.setAttribute('http.response.status_code', statusCode);
    }

    if (error !== undefined) {
      span.recordException(error instanceof Error ? error : new Error(String(error)));
      span.setStatus({ code: SpanStatusCode.ERROR, message: 'request failed' });
    } else if (statusCode === undefined) {
      span.setStatus({ code: SpanStatusCode.ERROR, message: 'no response' });
    } else if (statusCode >= 400) {
      // Unlike a server span, where 4xx is the caller's mistake, a 4xx here means *our*
      // outbound call failed. The distinction matters when reading a waterfall.
      span.setStatus({ code: SpanStatusCode.ERROR, message: `HTTP ${statusCode}` });
    }

    span.end();
  }
}

/** A malformed URL must not fail a request that is otherwise about to be made. */
function safeUrl(url: string | URL): URL | undefined {
  if (url instanceof URL) return url;
  try {
    return new URL(url);
  } catch {
    return undefined;
  }
}
